src/toucan/db.clj

```diff
@@ -746,14 +747,11 @@
   be deleted), or otherwise enforce preconditions before deleting (such as refusing to delete the object if
   something else depends on it).
 
-     (delete! Database :id 1)
-
-  NOTE: This function assumes objects have an `:id` column. There's an [open
-  issue](https://github.com/metabase/toucan/issues/3) to support objects that don't have one; until that is resolved,
-  you'll have to use `simple-delete!` instead when deleting objects with no `:id`."
+     (delete! Database :id 1)"
   {:style/indent 1}
   [model & conditions]
-  (let [model (resolve-model model)]
+  (let [model       (resolve-model model)
+        primary-key (models/primary-key model)]
     (doseq [object (apply select model conditions)]
       (models/pre-delete object)
-      (simple-delete! model :id (:id object)))))
+      (simple-delete! model primary-key (primary-key object)))))
```

Comment:
Is there any chance the primary key is not a keyword and possibly a string? If there is such a scenario, then `("primary-key" object)` will break.

Reply:
NVM... Question answered in function return type